# PheWeb2 API
This is an implementation of the data model and API for [PheWeb2](https://github.com/GaglianoTaliun-Lab/PheWeb2/tree/main) — an enhanced version of the original [PheWeb](https://github.com/statgen/pheweb) web-based tool for interactive querying, visualizing, and sharing summary-level results from genome-wide and phenome-wide association studies (GWAS/PheWAS), which offers intuitive and efficient support for stratified analysis results. PheWeb2 decouples the data model and API from the user interface (UI) to improve code maintenance and reusability and allow results querying by other external resources and applications. 
  
## 1. Install

> [!NOTE]
> The code was developed and tested with Python 3.12+ on Linux-based OS.

You can install PheWeb2 and all required dependencies within a virtual environment using the following steps:
1. Clone this repository:
```
git clone https://github.com/GaglianoTaliun-Lab/PheWeb2-API.git
cd PheWeb2-API
```
2. Create and activate Python virtual environment:
```
python -m venv .venv
source .venv/bin/activate
```
3. Install PheWeb2 Python package and its dependencies:
```
pip install -e .
```

## 2. Run using the example data
To familiarize yourself with PheWeb2, we recommend first trying to configure and run it with the provided example dataset by following the steps below.

1. Download and unarchive the example data (~13 GB):
```
wget https://objets.juno.calculquebec.ca/swift/v1/AUTH_290e6dcc5e264b34b401f54358bd4c54/pheweb_example_data/example_regenie.tar.gz
tar -xzvf example_regenie.tar.gz
```

2. Generate JSON file describing the phenotypes:
```
pheweb2 phenolist import-phenolist pheno-list-example.csv 
```

3. Ingest the example data into PheWeb2 (this can take some time):
```
pheweb2 process
```

4.  Launch PheWeb2 API endpoint which will be available at `http://127.0.0.1:9543`:
```
pheweb2-api
```

## 3. Run using your own data

### 3.1. Configuration file `config.py`
The self-documenting [config.py](config.py) configuration file includes all the variables that determine where PheWeb 2 API stores ingested GWAS data, how it processes this data, and how it serves it through HTTP. When performing GWAS data ingestion, you should focus on adjusting the variables in *SECTION A*, *SECTION B*, and *SECTION C* of this file.

### 3.2. Minimal GWAS summary statistics file

> [!NOTE]
> The current version of PheWeb 2 API has been tested using GWAS summary statistics output files generated by [Regenie](https://rgcgithub.github.io/regenie/).

> [!IMPORTANT]
> PheWeb 2 relies on the `TEST`/`test` column in GWAS summary statistics files to distinguish variant effects from interactions. While Regenie and [PLINK2](https://www.cog-genomics.org/plink/2.0/) GWAS ummary statistics files by default include the `TEST` column, which provides details on the statistical tests reported (e.g., *ADD*, *ADD-CONDTL*, *ADD-INT_SNPxVAR*, *ADDxCOVAR1*), other tools may not include this column. In such cases, we recommend adding a `TEST` column with the appropriate values to your GWAS ummary statistics files. Then, you can define the values representing each test in your GWAS ummary statistics files using `ASSOC_TEST_NAME` and `INTERACTION_TEST_NAME` variables in the [config.py](config.py) file.

The table below presents a minimal list of expected columns in GWAS summary statistics files. You can modify their names using the `FIELD_ALIASES` variable in the [config.py](config.py) file.

| column description | column name | allowed values              |
| ------------------ | ----------- | --------------------------- |
| Chromosome         | CHROM       | 1-22, 23/X                  |
| Position           | GENPOS      | Integer                     |
| Reference Allele   | ALLELE0     | Must match reference genome |
| Effect Allele      | ALLELE1     | Anything                    |
| Beta               | BETA        | Float                       |
| Standard Error     | SE          | Float                       |
| Log-10 P-value     | LOG10P      | Float                       |
| Statistical Test   | TEST        | Anything. Typical values: ADD, ADD-CONDTL, ADD-INT_SNPxVAR, ... |

### 3.3. Summary statistics exclusion criteria
During the ingestion of GWAS summary statistics, PheWeb 2 will apply the following exclusions:
- If a required field is null (i.e., any of the following: [’’, ‘.’, ‘NA’, ‘N/A’, ‘n/a’, ‘nan’, ‘-nan’, ‘NaN’, ‘-NaN’, ‘null’, ‘NULL’]), the variant will be discarded.
- If the minor allele frequency (MAF) of the variant falls below a specified threshold (defined by the `ASSOC_MIN_MAF` variable in the [config.py](config.py) file), the variant will be excluded.
- If the MAF of the variant is below a specified threshold (defined by the `INTERACTION_MIN_MAF` variable in the [config.py](config.py) file), the interaction effect for this variant will not be imported.
- When genotype imputation quality scores are available, if the imputation quality for a variant falls below a specified threshold (defined by the `MIN_IMP_QUALITY` variable in the [config.py](config.py) file), the variant will be excluded.

### 3.4. Genotype imputation qualities
In the preprocessing, you can filter out variants that have imputation quality lower than a customized threshold. To implement this, you can use specific field from the GWAS results (e.g., the INFO field in REGENIE output). You can specify the imputation quality field and threshold in the `config.py` by
```
MIN_IMP_QUALITY = # your imputation quality threshold
field_aliases = {
    ...
    # Add this field if you have imputation quality scores in the GWAS results
    # e.g., if the imputation quality field in the GWAS results is called 'INFO'
    "INFO": "imp_quality",
    ...
}
```
***
However, we strongly recommend you to access the imputation quality scores from external files, especially if you are using REGENIE to generate the GWAS results. We strongly recommend to use the VCF files (input of GWAS) as the source of the imputation quality scores.
***
You can set it up in the `config.py` by
```
MIN_IMP_QUALITY = # your imputation quality threshold
field_aliases = {
    ...
    # Start your custom field (e.g. imp_quality) with "file://" to load from a file (accept vcf files for imputation quality). Separate the file path the the field of interest by ','
    "file://path/file.vcf.gz,R2": "imp_quality",
    ...
}
```


## Preprocessing

 
### Imputation Quality Filtering
 
### Interaction Testing

If you wish to pre-process interaction results via Regenie output, you need to set 'interaction_aliases' to the value you gave regenie for the interaction testing and remap it to a value of your choice.

See sample_config.py for an example.

### Pheno-list{.csv, .json}

Like the `config.py` file, a pheno-list.json file MUST be in the base directory of the PheWeb-API before running any command.
- `PheWeb2.0-API/config.py`
- `PheWeb2.0-API/pheno-list.json`

We highly recommend creating a pheno-list.csv file, then converting to pheno-list.json. 


Here are the required and optional columns for the pheno-list.csv file:
 
| column description                                  | value         | allowed values                      | required? |
| --------------------------------------------------- | ------------- | ----------------------------------- | --------- |
| Phenotype Code                                      | phenocode     | string                              | true      |
| Phenotype Description                               | phenostring   | string                              | true      |
| Location of summary statistics                      | assoc_files   | string                              | true      |
| Number of tested samples / participants             | num_samples   | int                                 | false     |
| Number of tested cases                              | num_cases     | int                                 | false     |
| Number of tested controls                           | num_controls  | int                                 | false     |
| Category of trait                                   | category      | float                               | false     |
| Variable of Interaction Testing                     | interaction   | string                              | false     |
| Category of stratification (Can be more than one)   | interaction   | "stratification.*" (where *=string) | false     |

Refer to pheno-list-example.csv for an example.
Then with 'pheno-list.csv', run:

`pheweb phenolist import-phenolist "/path/to/pheno-list.csv"`

to create `pheno-list.json`.


To pre-process all files to properly work with the backend server, run:

`pheweb process`

To manually enable the autocomplete functionality of searching variant id or rsid, please run:

`pheweb generate-autocomplete-db`

to create `autocomplete.db`. Alternative option: the `autocomplete.db` will be created when you first run the api.

## Runtime Data
After pre-processing, these folders (with data) need to be present before running.

```
CLSA_PheWeb_data
└── generated_by_pheweb
    ├── manhattan
        ├── ... 
    ├── qq
        ├── ...
    ├── phenotypes.json 
    ├── top_hits.json 
    ├── ...
```

The paths to the runtime data needs to be specified in the config.py
```py
BASE_DIR = os.path.join(os.sep, 'var', 'local', 'CLSA_PheWeb_data', 'generated_by_pheweb')

PHENOTYPES_DIR = os.path.join(BASE_DIR)
...
```
please refer the `sample_config.py` for more examples.
 

## Running Backend Server
```
pheweb-run
```

get data through http://localhost:9099/PATH/TO/ROUTES/

e.g. to view phenotype table data, go to http://localhost:9099/phenotypes
 
## Documentation

To see all available API routes, visit http://localhost:9099/ (or the location of your backend API instance).

Documentation was created with [Flask-RESTX](https://flask-restx.readthedocs.io/en/latest/).


