# PheWeb2 API
This is an implementation of the data model and API for [PheWeb2](https://github.com/GaglianoTaliun-Lab/PheWeb2/tree/main) — an enhanced version of the original [PheWeb](https://github.com/statgen/pheweb) web-based tool for interactive querying, visualizing, and sharing summary-level results from genome-wide and phenome-wide association studies (GWAS/PheWAS), which offers intuitive and efficient support for stratified analysis results. PheWeb2 decouples the data model and API from the user interface (UI) to improve code maintenance and reusability and allow results querying by other external resources and applications. 
  
## 1. Install

> [!NOTE]
> The code was developed and tested with Python 3.12+ on Linux-based OS.

You can install PheWeb2 and all required dependencies within a virtual environment using the following steps:
1. Clone this repository:
```
git clone https://github.com/GaglianoTaliun-Lab/PheWeb2-API.git
cd PheWeb2-API
```
2. Create and activate Python virtual environment:
```
python -m venv .venv
source .venv/bin/activate
```
3. Install PheWeb2 Python package and its dependencies:
```
pip install -e .
```

## 2. Run using the example data
To familiarize yourself with PheWeb2, we recommend first trying to configure and run it with the provided example dataset by following the steps below.

1. Download and unarchive the example data (~13 GB):
```
wget https://objets.juno.calculquebec.ca/swift/v1/AUTH_290e6dcc5e264b34b401f54358bd4c54/pheweb_example_data/example_regenie.tar.gz
tar -xzvf example_regenie.tar.gz
```

2. Generate JSON file describing the phenotypes:
```
pheweb2 phenolist import-phenolist manifest-example.csv 
```

3. Ingest the example data into PheWeb2 (this can take some time):
```
pheweb2 process
```

4.  Launch PheWeb2 API endpoint which will be available at `http://127.0.0.1:9543`:
```
pheweb2-api
```

## 3. Run using your own data

### 3.1. Configuration file `config.py`
The self-documenting [config.py](config.py) configuration file includes all the variables that determine where PheWeb 2 API stores ingested GWAS data, how it processes this data, and how it serves it through HTTP. When performing GWAS data ingestion, you should focus on adjusting the variables in *SECTION A*, *SECTION B*, and *SECTION C* of this file:

- *SECTION A* of the [config.py](config.py) file lists the configuration variables that control the location of the processed GWAS summary statistics.
- *SECTION B* of the [config.py](config.py) file lists the configuration variables that control the versions of the external public databases such as dbSNP and GENCODE.
- *SECTION C* of the [config.py](config.py) file lists the configuration variables that control GWAS summary statistics ingestion.

### 3.2. Minimal GWAS summary statistics file

> [!NOTE]
> The current version of PheWeb 2 API has been tested using GWAS summary statistics output files generated by [Regenie](https://rgcgithub.github.io/regenie/).

> [!IMPORTANT]
> PheWeb 2 relies on the `TEST`/`test` column in GWAS summary statistics files to distinguish variant effects from interactions. While Regenie and [PLINK2](https://www.cog-genomics.org/plink/2.0/) GWAS ummary statistics files by default include the `TEST` column, which provides details on the statistical tests reported (e.g., *ADD*, *ADD-CONDTL*, *ADD-INT_SNPxVAR*, *ADDxCOVAR1*), other tools may not include this column. In such cases, we recommend adding a `TEST` column with the appropriate values to your GWAS ummary statistics files. Then, you can define the values representing each test in your GWAS ummary statistics files using `ASSOC_TEST_NAME` and `INTERACTION_TEST_NAME` variables in the [config.py](config.py) file.

The table below presents a minimal list of expected columns in GWAS summary statistics files. You can modify their names using the `FIELD_ALIASES` variable in the [config.py](config.py) file.

| column description | column name | allowed values              |
| ------------------ | ----------- | --------------------------- |
| Chromosome         | CHROM       | 1-22, 23/X                  |
| Position           | GENPOS      | Integer                     |
| Reference Allele   | ALLELE0     | Must match reference genome |
| Effect Allele      | ALLELE1     | Anything                    |
| Beta               | BETA        | Float                       |
| Standard Error     | SE          | Float                       |
| Log-10 P-value     | LOG10P      | Float                       |
| Statistical Test   | TEST        | Anything. Typical values: ADD, ADD-CONDTL, ADD-INT_SNPxVAR, ... |

### 3.3. Summary statistics exclusion criteria
During the ingestion of GWAS summary statistics, PheWeb 2 will apply the following exclusions:
- If a required field is null (i.e., any of the following: [’’, ‘.’, ‘NA’, ‘N/A’, ‘n/a’, ‘nan’, ‘-nan’, ‘NaN’, ‘-NaN’, ‘null’, ‘NULL’]), the variant will be discarded.
- If the minor allele frequency (MAF) of the variant falls below a specified threshold (defined by the `ASSOC_MIN_MAF` variable in the [config.py](config.py) file), the variant will be excluded.
- If the MAF of the variant is below a specified threshold (defined by the `INTERACTION_MIN_MAF` variable in the [config.py](config.py) file), the interaction effect for this variant will not be imported.
- When genotype imputation quality scores are available, if the imputation quality for a variant falls below a specified threshold (defined by the `MIN_IMP_QUALITY` variable in the [config.py](config.py) file), the variant will be excluded.

> [!NOTE]
> When filtering by imputation quality scores, the scores can be provided as a column inside the GWAS summary statistics file (e.g., the Regenie outputs imputation quality scores in the `INFO` column) or in the external indexed VCF files outputted by the genotype imputation software (e.g., [minimac4](https://github.com/statgen/Minimac4)). See an example inside the [config.py](config.py) file.

> [!IMPORTANT]
> The imputation quality scores recomputed by Regenie may differ from the scores outputted by the genotype imputation software, as Regenie uses dosages only for a subset of individuals included in GWAS who have phenotype and covariate data. When conducting stratified analyses, we recommend using the original genotype imputation quality scores computed by the imputation software.

### 3.4. Creating the Manifest file

To ingest GWAS summary statistics files into PheWeb2, you need to create the Manifest file. The Manifest file is a comma-separated (CSV) file that describes each GWAS summary statistics file you have. The [manifest-example.csv](manifest-example.csv) file serves as an example of the Manifest file, and the table below lists the required and optional columns.

| column description                                  | value         | allowed values                      | required? |
| --------------------------------------------------- | ------------- | ----------------------------------- | --------- |
| Phenotype Code                                      | phenocode     | string                              | true      |
| Phenotype Description                               | phenostring   | string                              | true      |
| Location of summary statistics                      | assoc_files   | string                              | true      |
| Number of tested samples / participants             | num_samples   | int                                 | true     |
| Number of tested cases                              | num_cases     | int                                 | false     |
| Number of tested controls                           | num_controls  | int                                 | false     |
| Category of trait                                   | category      | float                               | false     |
| Variable of Interaction Testing                     | interaction   | string                              | false     |
| Category of stratification (Can be more than one)   | interaction   | "stratification.*" (where *=string) | false     |


### 3.5. Ingesting GWAS summary statistics files
Once you have edited the [config.py](config.py) file and created the Manifest file, execute the following commands inside the root directory of the PheWeb2 API.

1. Import the manifest file into PheWeb2:
```
pheweb2 phenolist import-phenolist /path/to/manifest.csv
```
> [!NOTE]
> This command will create a `pheno-list.json` file in the root directory.

2. Ingest the GWAS summary statistics into PheWeb2:
```
pheweb2 process
```

> [!NOTE]
> This command is parallelized, but may still take some time depending on the size of your GWAS data.
> The ingested data will be by default stored inside the `generated_by_pheweb` folder, but this can be controlled using the `PHEWEB_DATA_DIR` variable inside the [config.py](config.py) file.
> After the data ingestion, you may archive your original GWAS summary statistics files, as PheWeb2 will no longer need them.
 

### 3.6. Running API server
```
pheweb2-api
```

get data through http://localhost:9099/PATH/TO/ROUTES/

e.g. to view phenotype table data, go to http://localhost:9099/phenotypes
 
## Documentation

To see all available API routes, visit http://localhost:9099/ (or the location of your backend API instance).

Documentation was created with [Flask-RESTX](https://flask-restx.readthedocs.io/en/latest/).


